### PR TITLE
📝 Clarify misleading "Required, can be `None`" section in `query-params-str-validations.md`

### PR DESCRIPTION
**Pull request by ChaitanyaSai-Meka**
The previous explanation suggested that a query parameter could be both required and accept a true `None` value.
However, since HTTP query parameters are transmitted as strings, clients cannot actually send `None` or `null`.

This update clarifies that behavior and adds an example showing how to manually handle `"None"` or empty strings as `None` inside a route.

### CODE REVIEW

**Comment by ChaitanyaSai-Meka** [comment]
Hi 👋,
This PR clarifies a small inaccuracy in the “Required, can be None” section of the Query Params tutorial.
It explains that query parameters can’t actually carry None and adds a short example showing how to handle special values like "None" or an empty string.

This is my first contribution to FastAPI 🎉 — really excited to start contributing to the project!

**Review by YuriiMotov — CHANGES_REQUESTED** [review]
@ChaitanyaSai-Meka, thanks for your interest and efforts!

Personally, I like the explanation!

We need to fix formatting (format `None`, `q=None`, etc.. with back-ticks), move code example to separate `.py` file in  `docs_src/query_params_str_validations/`, and cover it by test (tests for docs examples are in `tests/test_tutorial/query_params_str_validations/`)

File: docs_src/query_params_str_validations/tutorial006d_an_py310.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+from fastapi import FastAPI, Query
+from typing import Annotated
+
+app = FastAPI()
+
+@app.get("/items/")
+async def read_items(q: Annotated[str | None, Query()] = ...):
```
Comment (YuriiMotov) [inline docs_src/query_params_str_validations/tutorial006d_an_py310.py]:
```suggestion
async def read_items(q: Annotated[str | None, Query()]):
```
Ellipsis is not needed here

File: docs/en/docs/tutorial/query-params-str-validations.md [hunk 2]
```diff
@@ -271,23 +271,12 @@ For example, you might try:
 q: Annotated[str | None, Query(min_length=3)] = ...
 ```
 
-But this will still expect a string value, and if the client omits q or tries to send q=None, FastAPI will raise a validation error.
-In other words, None is not an acceptable runtime value for query parameters — only strings are.
+But this will still expect a **string** value, and if the client omits `q` or tries to send `q=None`, FastAPI will raise a validation error.
```
Comment (YuriiMotov) [inline docs/en/docs/tutorial/query-params-str-validations.md]:
```suggestion
But this will still expect a **string** value, and if the client omits `q` or tries to send `q=None`, FastAPI will raise a request validation error.
```

File: docs_src/query_params_str_validations/tutorial006d_an_py310.py (hunk 1 above)
Comment (YuriiMotov) [reply docs_src/query_params_str_validations/tutorial006d_an_py310.py]: Another thing is that this code changes `Query(min_length=3)` that is mentioned earlier to just `Query()`.
We should update previous code
````
```Python
q: Annotated[str | None, Query(min_length=3)] = ...
```
````
to
````
```Python
q: Annotated[str | None, Query()]
```
````
For consistency.

And then we can add something like "Stop, but what about `min_length=3`?" and introduce alternative approach with `BeforeValidator` that we can use **with Pydantic V2**:

```py

...
from pydantic import BeforeValidator

def nullable_str(val: str) -> Union[str, None]:
    if val in ("None", "", "null"):
        return None
    return val

async def read_items(q: Annotated[str | None, Query(min_length=3), BeforeValidator(nullable_str)]):
    return {"q": q}
```
(I haven't checked the code example, but should work)

**Comment by ChaitanyaSai-Meka** [comment]
Hi ,
I’ve addressed all requested changes:

Added inline code formatting for None, q, and q=None

Moved the example to docs_src/query_params_str_validations/tutorial006d_an_py310.py

Added corresponding test under tests/test_tutorial/test_query_params_str_validations/test_tutorial006d_an_py310.py

File: tests/test_tutorial/test_query_params_str_validations/test_tutorial006d_an_py310.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from fastapi.testclient import TestClient
+from docs_src.query_params_str_validations.tutorial006d_an_py310 import app
+
+client = TestClient(app)
+
+def test_read_items():
+    response = client.get("/items/", params={"q": "None"})
```
Comment (YuriiMotov) [inline tests/test_tutorial/test_query_params_str_validations/test_tutorial006d_an_py310.py]: We should parameterize this test to pass all possible values: `("None", "", "null")`, not just `"None"`.
And, I think we should add test with not-null parameter value

File: docs/en/docs/tutorial/query-params-str-validations.md [hunk 4]
```diff
@@ -271,23 +271,12 @@ For example, you might try:
...
-
-app = FastAPI()
-
-@app.get("/items/")
-async def read_items(q: Annotated[str | None, Query()] = ...):
-    if q in ("None", "", "null"):
-        q = None
-    return {"q": q}
-```
+{* ../../docs_src/query_params_str_validations/tutorial006d_an_py310.py hl[9] *}
```
Comment (YuriiMotov) [inline docs/en/docs/tutorial/query-params-str-validations.md]:
```suggestion
{* ../../docs_src/query_params_str_validations/tutorial006d_an_py310.py hl[9:11] *}
```

**Review by YuriiMotov — COMMENTED** [review]
Thanks!

A few more suggestions!)

File: docs_src/query_params_str_validations/tutorial006d_an_py310.py (hunk 1 above)
Comment (YuriiMotov) [reply docs_src/query_params_str_validations/tutorial006d_an_py310.py]: I can help with the last part (adding code example with `BeforeValidator`) if you need

**Comment by ChaitanyaSai-Meka** [comment]
Thanks a lot @YuriiMotov 🙏
I’ll apply your suggestions:

Remove ellipsis

Keep min_length=3 consistent

Add a BeforeValidator example for Pydantic V2

Expand test coverage to handle all possible values

Fix “request validation error” wording

Will update and push shortly

**Comment by ChaitanyaSai-Meka** [comment]
@YuriiMotov
Implemented all requested changes:

Removed ellipsis and kept min_length=3 consistent

Added BeforeValidator example for Pydantic v2

Expanded test coverage for ("None", "", "null", "hello")

Updated “request validation error” wording

Verified locally 
Ready for re-review

**Comment by ChaitanyaSai-Meka** [comment]
Hi @YuriiMotov 👋

I’ve confirmed that all requested changes have been implemented, 
rebased the branch on the latest `master`, and verified that local tests pass successfully ✅

It seems the CI jobs failed due to cached environment setup, not due to the changes themselves.  
Could you please re-run the workflows when convenient?

Thank you for reviewing and helping out!

**Comment by ChaitanyaSai-Meka** [comment]
Screenshot 2025-11-10 at 3 58 54 PM

File: docs_src/query_params_str_validations/tutorial006d_an_py310.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+from typing import Annotated, Optional, Union
```
Comment (YuriiMotov) [inline docs_src/query_params_str_validations/tutorial006d_an_py310.py]:
```suggestion
from typing import Optional, Union
from typing_extensions import Annotated
```

Otherwise it fails for Python 3.8

Comment (ChaitanyaSai-Meka) [reply docs_src/query_params_str_validations/tutorial006d_an_py310.py]: Thanks @YuriiMotov!
I’ve updated th import to use typing_extensions.Annotated for Python 3.8 compatibility ✅
Just pushed the fix , everything should be ready now!

File: tests/test_tutorial/test_query_params_str_validations/test_tutorial006d_an_py310.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import pytest
+from fastapi.testclient import TestClient
+
+from docs_src.query_params_str_validations.tutorial006d_an_py310 import app
+
+client = TestClient(app)
```
Comment (YuriiMotov) [inline tests/test_tutorial/test_query_params_str_validations/test_tutorial006d_an_py310.py]: We can't import it like this because this way tests will fail in the environment with Pydantic v1.
We should create a fixture like his:
```py
@pytest.fixture
def client():
    from docs_src.query_params_str_validations.tutorial006d_an_py310 import app
    yield TestClient(app)
```

And use it in `test_read_items`.
We should also mark `test_read_items` with `@needs_pydanticv2` to skip that test for env with Pydantic v1:


```
from .utils import needs_pydanticv2

...
@needs_pydanticv2
def test_read_items(q_value, expected, client: TestClient):
    ...
```

Comment (ChaitanyaSai-Meka) [reply tests/test_tutorial/test_query_params_str_validations/test_tutorial006d_an_py310.py]: Thanks @YuriiMotov!
I’ve rebased the branch and pushed the final test update (tests.utils import).
Verified locally, all works perfectly ✅

**Comment by YuriiMotov** [comment]
One last thing is missing: we need to clarify that this method (with `BeforeValidator`) will only work with Pydantic V2.

**Comment by ChaitanyaSai-Meka** [comment]
@YuriiMotov Added clarification that BeforeValidator is available only in Pydantic v2 ✅

File: docs/en/docs/tutorial/query-params-str-validations.md [hunk 7]
```diff
@@ -260,11 +260,25 @@ So, when you need to declare a value as required while using `Query`, you can si
...
+```
+
+But this will still expect a **string** value, and if the client omits `q` or tries to send `q=None`, FastAPI will raise a **request validation error**.
+In other words, `None` is not an acceptable runtime value for query parameters — only strings are.
+
+If you want to accept special values (like `"None"` or an empty string) and interpret them as `None` in your application, you can handle them manually in your function:
+
+{* ../../docs_src/query_params_str_validations/tutorial006d_an_py310.py hl[9,11] *}
+
+> **Note**: This example uses `BeforeValidator`, which is only available in **Pydantic v2**.
```
Comment (YuriiMotov) [inline docs/en/docs/tutorial/query-params-str-validations.md]:
```suggestion
/// note

This example uses `BeforeValidator`, which is only available in **Pydantic v2**.

///
```

File: docs/en/docs/tutorial/query-params-str-validations.md [hunk 8]
```diff
@@ -260,11 +260,25 @@ So, when you need to declare a value as required while using `Query`, you can si
...
+```Python
+q: Annotated[str | None, Query(min_length=3)] = ...
+```
+
+But this will still expect a **string** value, and if the client omits `q` or tries to send `q=None`, FastAPI will raise a **request validation error**.
+In other words, `None` is not an acceptable runtime value for query parameters — only strings are.
+
+If you want to accept special values (like `"None"` or an empty string) and interpret them as `None` in your application, you can handle them manually in your function:
+
+{* ../../docs_src/query_params_str_validations/tutorial006d_an_py310.py hl[9,11] *}
```
Comment (YuriiMotov) [inline docs/en/docs/tutorial/query-params-str-validations.md]:
```suggestion
{* ../../docs_src/query_params_str_validations/tutorial006d_an_py310.py hl[9:13 18] *}
```

**Review by YuriiMotov — APPROVED** [review]
LGTM!

@ChaitanyaSai-Meka, thanks!

**Comment by ChaitanyaSai-Meka** [comment]
Thanks @YuriiMotov! Really appreciate your review and guidance 🙌
Glad to have contributed to FastAPI!

File: docs/en/docs/tutorial/query-params-str-validations.md [hunk 9]
```diff
@@ -260,11 +260,29 @@ So, when you need to declare a value as required while using `Query`, you can si
...
+```Python
+q: Annotated[str | None, Query(min_length=3)] = ...
+```
+
+But this will still expect a **string** value, and if the client omits `q` or tries to send `q=None`, FastAPI will raise a **request validation error**.
+In other words, `None` is not an acceptable runtime value for query parameters — only strings are.
+
+If you want to accept special values (like `"None"` or an empty string) and interpret them as `None` in your application, you can handle them manually in your function:
+
+{* ../../docs_src/query_params_str_validations/tutorial006d_an_py310.py hl[9:13 18] *}
```
Comment (YuriiMotov) [inline docs/en/docs/tutorial/query-params-str-validations.md]:
```suggestion
{* ../../docs_src/query_params_str_validations/tutorial006d_an_py310.py hl[9:13,18] *}
```

File: docs/en/docs/tutorial/query-params-str-validations.md [hunk 10]
```diff
@@ -260,11 +260,29 @@ So, when you need to declare a value as required while using `Query`, you can si
...
+```Python
+q: Annotated[str | None, Query(min_length=3)] = ...
+```
+
+But this will still expect a **string** value, and if the client omits `q` or tries to send `q=None`, FastAPI will raise a **request validation error**.
+In other words, `None` is not an acceptable runtime value for query parameters — only strings are.
+
+If you want to accept special values (like `"None"` or an empty string) and interpret them as `None` in your application, you can handle them manually in your function:
+
+{* ../../docs_src/query_params_str_validations/tutorial006d_an_py310.py hl[9:13,18] *}
```
Comment (YuriiMotov) [inline docs/en/docs/tutorial/query-params-str-validations.md]:
```suggestion
{* ../../docs_src/query_params_str_validations/tutorial006d_an_py310.py hl[10:13,18] *}
```

**Comment by ChaitanyaSai-Meka** [comment]
Hi @YuriiMotov 👋
Just wanted to check if this PR is now queued for merge. No rush at all , just noticed that all checks have passed and it's been a few weeks, so I thought I'd follow up. Thanks again for the guidance!

**Comment by YuriiMotov** [comment]
> Hi @YuriiMotov 👋 Just wanted to check if this PR is now queued for merge. No rush at all , just noticed that all checks have passed and it's been a few weeks, so I thought I'd follow up. Thanks again for the guidance!

Hi!
Yes, we are waiting for final review by Sebastian

**Comment by ChaitanyaSai-Meka** [comment]
Got it, @YuriiMotov . thank you for the update!
I really appreciate your help and all the guidance throughout this PR.
Looking forward to the final merge.

**Comment by Shaan-alpha** [comment]
Thanks for clarifying and for continuing the work on this PR.

I had independently worked on the same docs behavior around required nullable query parameters in #15514, so I’ll follow this thread instead.

If any additional cleanup, tests, wording improvements, or review support would still be useful here, I’d be happy to help contribute.
